mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraImapUidDAO.java

```diff
@@ -0,0 +1,121 @@
+import org.apache.james.mailbox.cassandra.CassandraMessageId;
+import org.apache.james.mailbox.model.ComposedMessageId;
+
+import com.datastax.driver.core.PreparedStatement;
+import com.datastax.driver.core.ResultSet;
+import com.datastax.driver.core.Session;
+import com.datastax.driver.core.querybuilder.QueryBuilder;
+
+public class CassandraImapUidDAO {
+
+    private final CassandraAsyncExecutor cassandraAsyncExecutor;
+    private final PreparedStatement deleteStatement;
+    private final PreparedStatement insertStatement;
+    private final PreparedStatement selectAllStatement;
+    private final PreparedStatement selectStatement;
+
+    public CassandraImapUidDAO(Session session) {
+        this.cassandraAsyncExecutor = new CassandraAsyncExecutor(session);
+        this.deleteStatement = deleteStatement(session);
+        this.insertStatement = insertStatement(session);
+        this.selectAllStatement = selectAllStatement(session);
+        this.selectStatement = selectStatement(session);
+    }
+
+    private PreparedStatement deleteStatement(Session session) {
+        return session.prepare(QueryBuilder.delete()
+                .from(TABLE_NAME)
+                .where(eq(MESSAGE_ID, bindMarker(MESSAGE_ID)))
+                .and(eq(MAILBOX_ID, bindMarker(MAILBOX_ID))));
+    }
+
+    private PreparedStatement insertStatement(Session session) {
+        return session.prepare(insertInto(TABLE_NAME)
+                .value(MESSAGE_ID, bindMarker(MESSAGE_ID))
+                .value(MAILBOX_ID, bindMarker(MAILBOX_ID))
+                .value(IMAP_UID, bindMarker(IMAP_UID)));
+    }
+
+    private PreparedStatement selectAllStatement(Session session) {
+        return session.prepare(select(FIELDS)
+                .from(TABLE_NAME)
+                .where(eq(MESSAGE_ID, bindMarker(MESSAGE_ID))));
+    }
+
+    private PreparedStatement selectStatement(Session session) {
+        return session.prepare(select(FIELDS)
+                .from(TABLE_NAME)
+                .where(eq(MESSAGE_ID, bindMarker(MESSAGE_ID)))
+                .and(eq(MAILBOX_ID, bindMarker(MAILBOX_ID))));
+    }
+
+    public CompletableFuture<Void> delete(CassandraMessageId messageId, CassandraId mailboxId) {
+        return cassandraAsyncExecutor.executeVoid(deleteStatement.bind()
+                .setUUID(MESSAGE_ID, messageId.get())
+                .setUUID(MAILBOX_ID, mailboxId.asUuid()));
+    }
+
+    public CompletableFuture<Void> insert(CassandraMessageId messageId, CassandraId mailboxId, MessageUid uid) {
+        return cassandraAsyncExecutor.executeVoid(insertStatement.bind()
+                .setUUID(MESSAGE_ID, messageId.get())
+                .setUUID(MAILBOX_ID, mailboxId.asUuid())
+                .setLong(IMAP_UID, uid.asLong()));
+    }
+
+    public Stream<ComposedMessageId> retrieve(CassandraMessageId messageId, Optional<CassandraId> mailboxId) {
+        return CassandraUtils.convertToStream(selectStatement(messageId, mailboxId).join())
```

Comment:
join should be done outside the dao, you have to return a CompletableFuture<Stream<ComposedMessageId>>